Skip test_enqueued_local_size if non-uniform work-groups not supported

FranklandJack opened the pull request:
* `test_enqueued_local_size` will be skipped for OpenCL < 2.0. However,
non-uniform work-groups became optional in 3.0 and so check that they
are supported before running `test_enqueued_local_size` and otherwise
skip.

FranklandJack commented:
I have if defd out this code since it will cause compilation errors if `CL_VERSION_3_0` is not defined. I'm not sure whether the CTS is intended to work this way, or if it should always be compiled with the most recent OpenCL headers anyway since it used with the ICD loader?

alycm commented:
Given that non-uniform workgroups are listed on the set of things to break down into issues (https://github.com/KhronosGroup/OpenCL-CTS/issues/773) it would helpful to file an issue for the "Need to add checks for lack of Non-Uniform Work Groups support." item in the OpenCL 3.0 project.  And then tick the box.

That will help with tracking progress being made on the 3.0 CTS.

FranklandJack commented:
> Given that non-uniform workgroups are listed on the set of things to break down into issues (#773) it would helpful to file an issue for the "Need to add checks for lack of Non-Uniform Work Groups support." item in the OpenCL 3.0 project. And then tick the box.
> 
> That will help with tracking progress being made on the 3.0 CTS.

@alycm I have created an issue https://github.com/KhronosGroup/OpenCL-CTS/issues/791 and commened on #773, but I don't think I have edit permissions to tick the box.

gwawiork commented:
LGTM  but
I am not sure if these defines with CL_VERSION_3_0 is needed rather we check device OpenCL version reported like [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/generic_address_space/main.cpp#L82)
I would be nice added the same code [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/non_uniform_work_group/main.cpp#L41)

FranklandJack commented:
> LGTM but
> I am not sure if these defines with CL_VERSION_3_0 is needed rather we check device OpenCL version reported like [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/generic_address_space/main.cpp#L82)
> It would be nice to have the same code [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/non_uniform_work_group/main.cpp#L41)

Thank you for your feedback, I wasn't sure myself, but is it possible that the CTS could be compiled without `CL_VERSION_3_0` say, `CL_VERSION_1_2`, in which case `CL_DEVICE_NON_UNIFORM_WORK_GROUP_SUPPORT` will not be defined since it is conditionally defined [here](https://github.com/KhronosGroup/OpenCL-Headers/blob/master/CL/cl.h#L404)

kpet commented:
The CTS must always be compiled with the latest headers and all features enabled so this is a non-issue.

A revised version of #816 is likely to supersede this PR.

FranklandJack commented:
> The CTS must always be compiled with the latest headers and all features enabled so this is a non-issue.
> 
> A revised version of #816 is likely to supersede this PR.

Thanks for your feedback. In that case I will wait until #816 is merged and then close this PR.

FranklandJack commented:
I can confirm that in local testing #816 fixes the issue that this PR attempted to solve.

alycm commented on test_conformance/basic/test_enqueued_local_size.cpp:
```diff
@@ -74,6 +74,27 @@ test_enqueued_local_size(cl_device_id device, cl_context context, cl_command_que
+            test_error(err, "clGetDeviceInfo failed.");
+        }
+
+        if (false == areNonUniformWorkGroupsSupported)
```
```suggestion
        if (CL_FALSE == areNonUniformWorkGroupsSupported)
```

FranklandJack replied:
Thanks @alycm, have updated.

Jeremy-Kemp commented on test_conformance/basic/test_enqueued_local_size.cpp:
```diff
@@ -74,6 +74,27 @@ test_enqueued_local_size(cl_device_id device, cl_context context, cl_command_que
+                              &areNonUniformWorkGroupsSupported, nullptr);
+        if (CL_SUCCESS != err)
+        {
+            test_error(err, "clGetDeviceInfo failed.");
```
Should return `TEST_FAIL` here.

FranklandJack replied:
Good catch, I've replace this whole block with `test_error_ret(err, "clGetDeviceInfo failed.", TEST_FAILED)". Thanks for the feedback!

bashbaug commented on test_conformance/basic/test_enqueued_local_size.cpp:
```diff
@@ -74,6 +74,27 @@ test_enqueued_local_size(cl_device_id device, cl_context context, cl_command_que
+        {
+            log_info("Skipping since non-uniform work group sizes are not "
+                     "supported\n");
+            return TEST_SKIPPED_ITSELF;
```
This is OK but it's a pretty big hammer since it means that none of this test will be executed without support for non-uniform work groups.  Would it instead be preferable to execute the test with uniform work groups on an OpenCL 3.0 device without non-uniform work group support so there is at least some coverage for `get_enqueued_local_size`?

FranklandJack replied:
Thanks for the feedback @bashbaug. This is a really good point, I've updated the test slightly so that instead of skipping it clamps the global work dimensions such that the local work-group size divides the global size in the case that non-uniform work-groups are not supported. Let me know if this is the type of thing you had in mind?

FranklandJack commented:
>     * This test currently compiles kernels with `-cl-std=CL2.0`, but this may not be supported by all OpenCL 3.0 devices.  For now, I think we can pass `-cl-std=CL3.0` for OpenCL 3.0 devices, but eventually we may need more complicated logic here.

@bashbaug Thanks for your feedback! This is a good point, I created an issue to track this a while back #823, although that issue kinda got conflated with the fact that some tests which were optional for 3.0 aren't getting skipped.

I've updated the test to dynamically select `-cl-std=CL3.0` for a OpenCL-3.0 device and otherwise fall back on `-cl-std=CL2.0`.

FranklandJack commented:
> Latest changes look good, thanks!
> 
> When we're enqueuing uniform work groups it's probably slightly safer to align up than to align down, just to guarantee we launch at least one work group, but in practice this shouldn't matter.

Oops, didn't think about the case we could end up with a global size less than the local size, good catch! I've updated the rounding logic to round up instead, and updated the commit message since it had gone a bit stale.

bashbaug commented:
Merging - this has two approvals and I'm guessing @alycm is OK with these changes.
